### PR TITLE
fix(ts): parenthesized assert and assign

### LINKED ISSUE #12932: TypeScript: 'assert and assign' vs createParenthesizedExpressions
## Bug Report

**Current behavior**

`SyntaxError: Invalid parenthesized assignment pattern (1:0)`

**Input Code**

```js
(a as any) = null;
```

**Expected behavior**

no error

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**

```js
require("@babel/parser").parse("(a as any) = null", {
  createParenthesizedExpressions: true,
  plugins: ["typescript"],
});
```

- Babel version(s): 7.13.4
- Node/npm version: Node 14.15.0
- OS: Windows 10
- How you are using Babel: API

**Additional context**
Works fine without `createParenthesizedExpressions`.
Related PR: https://github.com/babel/babel/pull/7098

### PR DESCRIPTION
**Pull request by fedeci**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #12932 
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

We treat `ParenthesizedExpression` node as a `TSAsExpression` in `#toAssignable()` so that the function recurses one more time.

### CODE REVIEW

**Review by nicolo-ribaudo — COMMENTED** [review]
Can you add a test that with the TS plugin and `createParenthesizedExpression` `({}) = x` still throws?

**Comment by fedeci** [comment]
It broke that, what about adding a check like this to the superclass' `toAssignable`
```js
isParenthesizedAssignable(node) {
  const type = node.type
  return (
    type === "TSParameterProperty" || 
    type === "TSAsExpression" || 
    ... || 
    super.isParenthesizedAssignable(node)
  )
}
```

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 1]
```diff
@@ -2680,6 +2680,7 @@ export default (superClass: Class<Parser>): Class<Parser> =>
           return super.toAssignable(this.typeCastToParameter(node), isLHS);
         case "TSParameterProperty":
           return super.toAssignable(node, isLHS);
+        case "ParenthesizedExpression":
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/plugins/typescript/index.js:2683]: Would doing this instead work?
```suggestion
        case "ParenthesizedExpression":
          if (node.expression.type === "TSAsExpression" || node.expression.type === "ParenthesizedExpression") {
            node.expression = this.toAssignable(node.expression, isLHS);
            return node;
          } else {
            return super.toAssignable(node, isLHS);
          }
```

Comment (fedeci) [reply packages/babel-parser/src/plugins/typescript/index.js:2683]: It does!

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 2]
```diff
@@ -2680,6 +2680,16 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
+          if (
+            node.expression.type === "TSAsExpression" ||
+            node.expression.type === "ParenthesizedExpression"
```
Comment (thorn0) [inline packages/babel-parser/src/plugins/typescript/index.js:2686]: I guess `TSNonNullExpression` and `TSTypeAssertion` should be here too

Comment (fedeci) [reply packages/babel-parser/src/plugins/typescript/index.js:2686]: Yes, I tried running some tests on those, but they work like magic😁

**Review by JLHwung — COMMENTED** [review]
Like @thorn0 said, we still need to check for `TSNonNullExpression` and `TSTypeAssertion`. It the test is passing maybe there isn't a test which enables `createParenthesizedExpressions`. Clearly Babel parser 7.12 fails: https://astexplorer.net/#/gist/910487b331a19aac83c5489fa4203e66/4180181cff60d53a164ab16afb82a8e6e7d5698c

The error is thrown from https://github.com/babel/babel/blob/efdca0140938de7606a42b3f5a04bc4f2513c7a9/packages/babel-parser/src/parser/lval.js#L91, which assumes that parenthesized assignment can be either identifier / member expression. This is not the cases for TypeScript, where `TSNonNullExpression`, `TSAsExpression` and `TSTypeAssertion` plays a role of transparent expression wrapper, which should have been unwrapped before check. Consider handle this as we did in 

https://github.com/babel/babel/blob/efdca0140938de7606a42b3f5a04bc4f2513c7a9/packages/babel-helper-skip-transparent-expression-wrappers/src/index.js#L11-L18

I think we can also simplifies how we handle `toAssignable` in the typescript plugin using this helper.

**Comment by fedeci** [comment]
> ```js
> 
> isParenthesizedAssignable(node) {
> 
>   const type = node.type
> 
>   return (
> 
>     type === "TSParameterProperty" || 
> 
>     type === "TSAsExpression" || 
> 
>     ... || 
> 
>     super.isParenthesizedAssignable(node)
> 
>   )
> 
> }
> ```
Maybe I got confused and I tested it with this code which does almost the same of that helper.

**Comment by JLHwung** [comment]
Can you add following test cases? 
```ts
(a!) = null;
(<string>a) = null;
```

Like Georgii mentioned before, they should not throw.

**Review by JLHwung — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
Thanks!

**Review by nicolo-ribaudo — COMMENTED** [review]
I'll merge this in 15 mins, as soon as the release CI job on `main` finishes

**Review by nicolo-ribaudo — APPROVED** [review]

**Pull request merged**